pkg/expansion/system_test.go

```diff
@@ -0,0 +1,829 @@
+			},
+			want: []*unstructured.Unstructured{loadFixture(fixtures.PodImagePullMutate, t)},
+		},
+		{
+			name:      "2 mutators, 2 matches, basic deployment expands pod",
+			generator: loadFixture(fixtures.DeploymentNginx, t),
+			ns:        &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default"}},
+			mutators: []types.Mutator{
+				loadAssign(fixtures.AssignPullImage, t),
+				loadAssignMeta(fixtures.AssignMetaAnnotatePod, t),
+			},
+			templates: []*expansionunversioned.ExpansionTemplate{
+				loadTemplate(fixtures.TempExpDeploymentExpandsPods, t),
+			},
+			want: []*unstructured.Unstructured{loadFixture(fixtures.PodImagePullMutateAnnotated, t)},
+		},
+		{
+			name:      "custom CR with 2 different resultant kinds, with mutators",
+			generator: loadFixture(fixtures.GeneratorCat, t),
+			ns:        &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default"}},
+			mutators: []types.Mutator{
+				loadAssign(fixtures.AssignKittenAge, t),
+				loadAssignMeta(fixtures.AssignMetaAnnotatePurr, t),
+				loadAssignMeta(fixtures.AssignMetaAnnotateKitten, t),
+			},
+			templates: []*expansionunversioned.ExpansionTemplate{
+				loadTemplate(fixtures.TemplateCatExpandsKitten, t),
+				loadTemplate(fixtures.TemplateCatExpandsPurr, t),
+			},
+			want: []*unstructured.Unstructured{
+				loadFixture(fixtures.ResultantKitten, t),
+				loadFixture(fixtures.ResultantPurr, t),
+			},
+		},
+		{
+			name:      "custom CR with 2 different resultant kinds, with mutators and non-matching configs",
+			generator: loadFixture(fixtures.GeneratorCat, t),
+			ns:        &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "default"}},
+			mutators: []types.Mutator{
+				loadAssign(fixtures.AssignKittenAge, t),
+				loadAssignMeta(fixtures.AssignMetaAnnotatePurr, t),
+				loadAssignMeta(fixtures.AssignMetaAnnotateKitten, t),
+				loadAssign(fixtures.AssignPullImage, t), // should not match
+			},
+			templates: []*expansionunversioned.ExpansionTemplate{
+				loadTemplate(fixtures.TemplateCatExpandsKitten, t),
+				loadTemplate(fixtures.TemplateCatExpandsPurr, t),
+				loadTemplate(fixtures.TempExpDeploymentExpandsPods, t), // should not match
+			},
+			want: []*unstructured.Unstructured{
+				loadFixture(fixtures.ResultantKitten, t),
+				loadFixture(fixtures.ResultantPurr, t),
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			mutSystem := mutation.NewSystem(mutation.SystemOpts{})
+			for _, m := range tc.mutators {
+				if err := mutSystem.Upsert(m); err != nil {
+					t.Fatalf("error upserting mutator: %s", err)
+				}
+			}
+
+			expSystem := NewSystem(mutSystem)
+			for _, te := range tc.templates {
+				if err := expSystem.UpsertTemplate(te); err != nil {
+					t.Fatalf("error upserting template: %s", err)
+				}
+			}
+
+			base := &types.Mutable{
+				Object:    tc.generator,
+				Namespace: tc.ns,
+				Username:  "unit-test",
+				Source:    types.SourceTypeGenerated,
```

Comment:
Is there a test for SourceOriginal and All?

Reply:
Great catch! I added tests for mutators with "Original", "All", and "" (which defaults to "All)